### PR TITLE
feat: Add timer and error messages in ui with unit tests

### PR DESCRIPTION
**Pull request by yksolanki9**
Clickup - https://app.clickup.com/t/85zt3yxe6

UI Preview [Changed timer to 5 sec only for video]:
https://github.com/fylein/fyle-mobile-app/assets/39493102/2a3b2ed2-37ae-444d-9a60-8086ca50e7b8

Code Coverage:
Screenshot 2023-05-24 at 7 34 07 PM

Screenshot 2023-05-25 at 9 06 05 AM

### CODE REVIEW

File: src/app/shared/directive/form-button-validation.directive.ts [hunk 1]
```diff
@@ -42,7 +44,13 @@ export class FormButtonValidationDirective implements OnInit, OnChanges {
   constructor(private elementRef: ElementRef) {}
 
   ngOnChanges(changes: SimpleChanges) {
-    this.onLoading(this.loading);
+    if (this.disabled) {
```
Comment (yksolanki9) [inline src/app/shared/directive/form-button-validation.directive.ts:47]: The directive was overwriting the disabled state of the cta, so added a check here to make sure the cta is always disabled

File: src/app/fyle/my-profile/verify-number-popover/verify-number-popover.component.html [hunk 2]
```diff
@@ -1,7 +1,7 @@
 <ion-header mode="md">
-  <ion-toolbar mode="md" class="fy-modal-toolbar verify-number-popover--toolbar">
```
Comment (yksolanki9) [inline src/app/fyle/my-profile/verify-number-popover/verify-number-popover.component.html:2]: Fixed class names to be as per BEM

File: src/app/fyle/my-profile/verify-number-popover/verify-number-popover.component.ts [hunk 3]
```diff
@@ -76,7 +106,34 @@ export class VerifyNumberPopoverComponent implements OnInit, AfterViewInit {
       .pipe(finalize(() => (this.verifyingOtp = false)))
       .subscribe({
         complete: () => this.popoverController.dismiss({ action: 'SUCCESS' }),
-        error: () => (this.error = 'Incorrect mobile number or OTP. Please try again.'),
+        error: () => this.setError('INVALID_OTP'),
       });
   }
+
+  setError(error: ErrorType, attemptsLeft = 5) {
+    const errorMapping = {
```
Comment (Dimple16) [inline src/app/fyle/my-profile/verify-number-popover/verify-number-popover.component.ts:114]: let's move the error mapping to enum?

File: src/app/fyle/my-profile/verify-number-popover/verify-number-popover.component.spec.ts [hunk 4]
```diff
@@ -179,12 +267,66 @@ describe('VerifyNumberPopoverComponent', () => {
...
       ) as HTMLSpanElement;
 
       expect(mobileNumberVerificationService.verifyOtp).not.toHaveBeenCalled();
       expect(popoverController.dismiss).not.toHaveBeenCalled();
       expect(getTextContent(errorElement)).toEqual(component.error);
     });
   });
+
+  it('setError(): should set correct error messages', () => {
+    const errorMappings = [
```
Comment (Dimple16) [inline src/app/fyle/my-profile/verify-number-popover/verify-number-popover.component.spec.ts:280]: add this to mock data file and create a model

**Review by Dimple16 — CHANGES_REQUESTED** [review]
added comments

File: src/app/fyle/my-profile/verify-number-popover/verify-number-popover.component.ts [hunk 5]
```diff
@@ -53,14 +65,32 @@ export class VerifyNumberPopoverComponent implements OnInit, AfterViewInit {
...
+            this.setError('ATTEMPTS_LEFT', attemptsLeft);
+            this.startTimer();
+          } else {
+            this.setError('LIMIT_REACHED');
+            this.disableResendOtp = true;
+          }
+        },
+        error: (err) => {
+          if (err.status === 400) {
+            const errorMessage = err.error.message.toLowerCase();
```
[the remark this answers is not in the document]
Comment (yksolanki9) [reply src/app/fyle/my-profile/verify-number-popover/verify-number-popover.component.ts:86]: Done

File: src/app/fyle/my-profile/verify-number-popover/verify-number-popover.component.spec.ts (hunk 4 above)
Comment (yksolanki9) [reply src/app/fyle/my-profile/verify-number-popover/verify-number-popover.component.spec.ts:280]: This array just contains the expected input/output for tests and is not being used anywhere in ts code, so I've created the model inside the data file itself

File: src/app/core/mock-data/error-mapping-for-verify-number-popover.data.ts [hunk 6]
```diff
@@ -0,0 +1,36 @@
+import { ErrorType } from 'src/app/fyle/my-profile/verify-number-popover/error-type.model';
+
+type ErrorMapping = {
```
Comment (yksolanki9) [inline src/app/core/mock-data/error-mapping-for-verify-number-popover.data.ts:3]: This array just contains the expected input/output for tests and is not being used anywhere in ts code, so I've created the model inside the data file itself

**Review by Dimple16 — APPROVED** [review]

**Pull request merged**